core: unable to comment on other org

logs:
```
2024/01/20 13:53:21.907505 [ERROR] github.go:118: retrying after error: POST https://api.github.com/repos/qiniu/reviewbot/pulls/21/comments: 403 Resource not accessible by integration []
```

--- a/deploy/reviewbot.yaml
+++ b/deploy/reviewbot.yaml
@@ -24,7 +24,6 @@ spec:
         - -webhook-secret=$(GITHUB_WEBHOOK_SECRET)
         - -config=/etc/config/config.yaml
         - -app-id=$(GITHUB_APP_ID)
-        - -app-installation-id=$(GITHUB_APP_INSTALLATION_ID)
         - -app-private-key=/secrets/github_app_key
         env:
         - name: GITHUB_WEBHOOK_SECRET
@@ -36,11 +35,6 @@ spec:
           valueFrom:
             secretKeyRef:
               key: app-id
-              name: github-app
-        - name: GITHUB_APP_INSTALLATION_ID
-          valueFrom:
-            secretKeyRef:
-              key: app-installation-id
               name: github-app                                
         image: aslan-spock-register.qiniu.io/qa/reviewbot:latest # use your own built image instead
         imagePullPolicy: Always

--- a/github.go
+++ b/github.go
@@ -1,12 +1,12 @@
 /*
  Copyright 2024 Qiniu Cloud (qiniu.com).
- 
+
  Licensed under the Apache License, Version 2.0 (the "License");
  you may not use this file except in compliance with the License.
  You may obtain a copy of the License at
- 
+
      http://www.apache.org/licenses/LICENSE-2.0
- 
+
  Unless required by applicable law or agreed to in writing, software
  distributed under the License is distributed on an "AS IS" BASIS,
  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
@@ -23,13 +23,13 @@ import (
 	"strings"
 	"time"
 
-	"github.com/reviewbot/internal/linters"
 	"github.com/google/go-github/v57/github"
 	"github.com/qiniu/x/log"
+	"github.com/reviewbot/internal/linters"
 )
 
 // ListFiles lists all files for the specified pull request.
-func (s *Server) ListPullRequestsFiles(ctx context.Context, owner string, repo string, number int) ([]*github.CommitFile, *github.Response, error) {
+func ListPullRequestsFiles(ctx context.Context, gc *github.Client, owner string, repo string, number int) ([]*github.CommitFile, *github.Response, error) {
 	opts := github.ListOptions{
 		PerPage: 100,
 	}
@@ -37,7 +37,7 @@ func (s *Server) ListPullRequestsFiles(ctx context.Context, owner string, repo s
 	var pullRequestAffectedFiles []*github.CommitFile
 
 	for {
-		files, response, err := s.gc.PullRequests.ListFiles(ctx, owner, repo, number, &opts)
+		files, response, err := gc.PullRequests.ListFiles(ctx, owner, repo, number, &opts)
 		if err != nil {
 			return nil, nil, err
 		}
@@ -52,10 +52,10 @@ func (s *Server) ListPullRequestsFiles(ctx context.Context, owner string, repo s
 	}
 }
 
-func (s *Server) PostPullReviewCommentsWithRetry(ctx context.Context, owner string, repo string, number int, comments []*github.PullRequestComment) error {
+func PostPullReviewCommentsWithRetry(ctx context.Context, gc *github.Client, owner string, repo string, number int, comments []*github.PullRequestComment) error {
 	var existedComments []*github.PullRequestComment
 	err := retryWithBackoff(ctx, func() error {
-		originalComments, resp, err := s.gc.PullRequests.ListComments(ctx, owner, repo, number, nil)
+		originalComments, resp, err := gc.PullRequests.ListComments(ctx, owner, repo, number, nil)
 		if err != nil {
 			return err
 		}
@@ -88,7 +88,7 @@ func (s *Server) PostPullReviewCommentsWithRetry(ctx context.Context, owner stri
 		}
 
 		err := retryWithBackoff(ctx, func() error {
-			_, resp, err := s.gc.PullRequests.CreateComment(ctx, owner, repo, number, comment)
+			_, resp, err := gc.PullRequests.CreateComment(ctx, owner, repo, number, comment)
 			if err != nil {
 				return err
 			}

--- a/main.go
+++ b/main.go
@@ -1,12 +1,12 @@
 /*
  Copyright 2024 Qiniu Cloud (qiniu.com).
- 
+
  Licensed under the Apache License, Version 2.0 (the "License");
  you may not use this file except in compliance with the License.
  You may obtain a copy of the License at
- 
+
      http://www.apache.org/licenses/LICENSE-2.0
- 
+
  Unless required by applicable law or agreed to in writing, software
  distributed under the License is distributed on an "AS IS" BASIS,
  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
@@ -23,11 +23,9 @@ import (
 	"net/http"
 	"os"
 
-	"github.com/bradleyfalzon/ghinstallation/v2"
-	"github.com/reviewbot/config"
 	"github.com/google/go-github/v57/github"
-	"github.com/gregjones/httpcache"
 	"github.com/qiniu/x/log"
+	"github.com/reviewbot/config"
 	"github.com/sirupsen/logrus"
 	gitv2 "k8s.io/test-infra/prow/git/v2"
 
@@ -56,8 +54,8 @@ func (o options) Validate() error {
 		return errors.New("either access-token or github app information should be provided")
 	}
 
-	if o.appID != 0 && o.installationID == 0 {
-		return errors.New("app-installation-id is required when using github app")
+	if o.appID != 0 && o.appPrivateKey == "" {
+		return errors.New("app-private-key is required when using github app")
 	}
 
 	if o.webhookSecret == "" {
@@ -93,18 +91,6 @@ func main() {
 	log.SetFlags(log.LstdFlags | log.Lshortfile | log.Llevel)
 	log.SetOutputLevel(o.logLevel)
 
-	var gc *github.Client
-	if o.appID != 0 {
-		tr, err := ghinstallation.NewKeyFromFile(httpcache.NewMemoryCacheTransport(), o.appID, o.installationID, o.appPrivateKey)
-		if err != nil {
-			log.Fatalf("failed to create github app transport: %v", err)
-		}
-		gc = github.NewClient(&http.Client{Transport: tr})
-	} else {
-		gc = github.NewClient(nil)
-		gc.WithAuthToken(o.accessToken)
-	}
-
 	if o.codeCacheDir != "" {
 		if err := os.MkdirAll(o.codeCacheDir, 0755); err != nil {
 			log.Fatalf("failed to create code cache dir: %v", err)
@@ -129,10 +115,12 @@ func main() {
 	}
 
 	s := &Server{
-		gc:               gc,
 		webhookSecret:    []byte(o.webhookSecret),
 		gitClientFactory: v2,
 		config:           cfg,
+		accessToken:      o.accessToken,
+		appID:            o.appID,
+		appPrivateKey:    o.appPrivateKey,
 	}
 
 	mux := http.NewServeMux()

--- a/server.go
+++ b/server.go
@@ -22,7 +22,10 @@ import (
 	"net/http"
 	"path/filepath"
 
+	"github.com/bradleyfalzon/ghinstallation/v2"
 	"github.com/google/go-github/v57/github"
+	"github.com/gregjones/httpcache"
+	"github.com/qiniu/x/log"
 	"github.com/qiniu/x/xlog"
 	"github.com/reviewbot/config"
 	"github.com/reviewbot/internal/linters"
@@ -30,11 +33,16 @@ import (
 )
 
 type Server struct {
-	gc               *github.Client
 	gitClientFactory gitv2.ClientFactory
 	config           config.Config
 
 	webhookSecret []byte
+
+	// support developer access token model
+	accessToken string
+	// support github app model
+	appID         int64
+	appPrivateKey string
 }
 
 func (s *Server) ServeHTTP(w http.ResponseWriter, r *http.Request) {
@@ -83,9 +91,10 @@ func (s *Server) handle(log *xlog.Logger, ctx context.Context, event *github.Pul
 	num := event.GetPullRequest().GetNumber()
 	org := event.GetRepo().GetOwner().GetLogin()
 	repo := event.GetRepo().GetName()
-	log.Infof("processing pull request %d, (%v/%v)\n", num, org, repo)
+	installationID := event.GetInstallation().GetID()
+	log.Infof("processing pull request %d, (%v/%v), installationID: %d\n", num, org, repo, installationID)
 
-	pullRequestAffectedFiles, response, err := s.ListPullRequestsFiles(ctx, org, repo, num)
+	pullRequestAffectedFiles, response, err := ListPullRequestsFiles(ctx, s.GithubClient(installationID), org, repo, num)
 	if err != nil {
 		return err
 	}
@@ -142,7 +151,7 @@ func (s *Server) handle(log *xlog.Logger, ctx context.Context, event *github.Pul
 		}
 
 		log.Infof("%s found valid %d comments related to this PR %d (%s) \n", name, len(comments), num, org+"/"+repo)
-		if err := s.PostPullReviewCommentsWithRetry(ctx, org, repo, num, comments); err != nil {
+		if err := PostPullReviewCommentsWithRetry(ctx, s.GithubClient(installationID), org, repo, num, comments); err != nil {
 			log.Errorf("failed to post comments: %v", err)
 			return err
 		}
@@ -163,7 +172,7 @@ func (s *Server) handle(log *xlog.Logger, ctx context.Context, event *github.Pul
 
 		log.Infof("running %s on repo %v with config %v", name, fmt.Sprintf("%s/%s", org, repo), lingerConfig)
 
-		agent := linters.NewAgent(s.gc, s.gitClientFactory, s.config)
+		agent := linters.NewAgent(s.GithubClient(installationID), s.gitClientFactory, s.config)
 		if err := fn(log, lingerConfig, agent, *event); err != nil {
 			log.Errorf("failed to run linter: %v", err)
 			return err
@@ -171,4 +180,26 @@ func (s *Server) handle(log *xlog.Logger, ctx context.Context, event *github.Pul
 	}
 
 	return nil
+}
+
+func (s *Server) githubAppClient(installationID int64) *github.Client {
+	tr, err := ghinstallation.NewKeyFromFile(httpcache.NewMemoryCacheTransport(), s.appID, installationID, s.appPrivateKey)
+	if err != nil {
+		log.Fatalf("failed to create github app transport: %v", err)
+	}
+	return github.NewClient(&http.Client{Transport: tr})
+}
+
+func (s *Server) githubAccessTokenClient() *github.Client {
+	gc := github.NewClient(httpcache.NewMemoryCacheTransport().Client())
+	gc.WithAuthToken(s.accessToken)
+	return gc
+}
+
+// GithubClient returns a github client
+func (s *Server) GithubClient(installationID int64) *github.Client {
+	if s.accessToken != "" {
+		return s.githubAccessTokenClient()
+	}
+	return s.githubAppClient(installationID)
 }

